### PR TITLE
rewrite: stream insert batch

### PR DESCRIPTION
**Pull request by Nateiru**
rewrite stream insert batch

### CODE REVIEW

File: src/stream_inserter.cpp [hunk 1]
```diff
@@ -13,34 +13,76 @@
...
+        cv.wait(lk, [this]{
+            return !this->buffer.empty() || !this->scheduler_thread_is_running;
+        });
+        if (buffer.empty() && !scheduler_thread_is_running) {
+            break;
+        }
+
+        if (!buffer.empty()) {
+            size_t batch_byte = 0;
+            #define BATCH_BYTE_LIMIT 2981328
```
Comment (v0y4g3r) [inline src/stream_inserter.cpp]: Maybe lift this constant to the start of file

File: src/stream_inserter.cpp [hunk 2]
```diff
@@ -13,34 +13,76 @@
...
+            InsertRequests insert_requests;
+            while (!buffer.empty() && batch_byte < BATCH_BYTE_LIMIT) {
+                if (batch_byte + buffer.front().ByteSizeLong() > BATCH_BYTE_LIMIT) {
+                    break;
+                } 
+                auto insert_request = std::move(buffer.front());
```
Comment (v0y4g3r) [inline src/stream_inserter.cpp:65]: Can we merge `buffer::front` with `buffer::pop`?
```suggestion
                auto insert_request = buffer.pop();
```

File: src/stream_inserter.cpp [hunk 3]
```diff
@@ -13,34 +13,76 @@
...
+                buffer.pop();
+                batch_byte += insert_request.ByteSizeLong();
+                insert_requests.add_inserts()->Swap(&insert_request);
+            }
+            lk.unlock();
+            RequestHeader request_header;
```
Comment (v0y4g3r) [inline src/stream_inserter.cpp:71]: Can we reuse the header instead of creating a new instance every time?

File: src/stream_inserter.h [hunk 4]
```diff
@@ -12,46 +12,87 @@
...
+    void Write(InsertRequest insert_request);
+
+    void WriteBatch(std::vector<InsertRequest> insert_request_vec);
 
-    bool WriteDone() { return writer->WritesDone();}
+    bool Send(const GreptimeRequest &greptime_request);
+
+    void RunHandleRequest();
+
+    bool WriteDone() {
```
Comment (v0y4g3r) [inline src/stream_inserter.h:80]: What's the difference between `WriteDone` and `Finish`? Can we explicitly doc these two methods?

File: src/stream_inserter.cpp (hunk 2 above)
Comment (Nateiru) [reply src/stream_inserter.cpp:65]: std::queue::pop doesn't return value: [Why doesn't std::queue::pop return value.?](https://stackoverflow.com/questions/25035691/why-doesnt-stdqueuepop-return-value)

File: src/stream_inserter.cpp (hunk 3 above)
Comment (Nateiru) [reply src/stream_inserter.cpp:71]: The header becomes null after swap. If we consider reusing it, we may need to use copy function, which has little difference in effect. And I think it's cleaner to create a new instance every time.

**Review by v0y4g3r — APPROVED** [review]
LGTM

**Pull request merged**